velox/functions/prestosql/ArrayPosition.cpp

```diff
@@ -0,0 +1,339 @@
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    FlatVector<int64_t>& flatResult) {
+  applyComplexTypeFirstMatch(
+      rows, arrayDecoded, elementsDecoded, searchDecoded, flatResult);
+}
+
+template <TypeKind kind>
+void applyTypedWithInstance(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    const DecodedVector& instanceDecoded,
+    FlatVector<int64_t>& flatResult) {
+  using T = typename TypeTraits<kind>::NativeType;
+
+  auto baseArray = arrayDecoded.base()->as<ArrayVector>();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = arrayDecoded.indices();
+
+  rows.applyToSelected([&](auto row) {
+    auto size = rawSizes[indices[row]];
+    auto offset = rawOffsets[indices[row]];
+    auto search = searchDecoded.valueAt<T>(row);
+
+    auto instance = instanceDecoded.valueAt<int64_t>(row);
+    VELOX_USER_CHECK_NE(
+        instance,
+        0,
+        "array_position cannot take a 0-valued instance argument.");
+
+    int startIndex = instance > 0 ? 0 : size - 1;
+    int endIndex = instance > 0 ? size : -1;
+    int step = instance > 0 ? 1 : -1;
+    instance = std::abs(instance);
+
+    for (auto i = startIndex; i != endIndex; i += step) {
+      if (!elementsDecoded.isNullAt(offset + i) &&
+          elementsDecoded.valueAt<T>(offset + i) == search) {
+        --instance;
+        if (instance == 0) {
+          flatResult.set(row, i + 1);
+          return;
+        }
+      }
+    }
+    flatResult.set(row, 0);
+  });
+}
+
+void applyComplexTypeWithInstance(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    const DecodedVector& instanceDecoded,
+    FlatVector<int64_t>& flatResult) {
+  auto baseArray = arrayDecoded.base()->as<ArrayVector>();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = arrayDecoded.indices();
+
+  auto elementsBase = elementsDecoded.base();
+
+  auto searchBase = searchDecoded.base();
+  auto searchIndices = searchDecoded.indices();
+
+  rows.applyToSelected([&](auto row) {
+    auto size = rawSizes[indices[row]];
+    auto offset = rawOffsets[indices[row]];
+    auto searchIndex = searchIndices[row];
+
+    auto instance = instanceDecoded.valueAt<int64_t>(row);
+    VELOX_USER_CHECK_NE(
+        instance,
+        0,
+        "array_position cannot take a 0-valued instance argument.");
+
+    int startIndex = instance > 0 ? 0 : size - 1;
+    int endIndex = instance > 0 ? size : -1;
+    int step = instance > 0 ? 1 : -1;
+    instance = std::abs(instance);
+
+    for (auto i = startIndex; i != endIndex; i += step) {
+      if (!elementsBase->isNullAt(offset + i) &&
+          elementsBase->equalValueAt(searchBase, offset + i, searchIndex)) {
```

Comment:
Nit: can we use equalValueAt to be also used for primitive types and thus allow us to have same function for complex and primitive types ?

Reply:
Good question. I think one difference between the code for primitive types and complex types is that we compare values in `DecodedVector::data_` when the type is primitive, while we compare values in `DecodedVector::baseVector_` when the type is complex. So I think `equalValueAt` cannot be used for primitive types (because `DecodedVector::data_` is a `const void*`).